packages/purpose-process/src/services/readModelService.ts

```diff
@@ -182,27 +183,29 @@ async function buildGetPurposesAggregation(
           )
       : [];
 
-  const eservicesIdsFilter: ReadModelFilter<Purpose> =
-    /**
-     * In case both producersIds and eservicesIds filters are present,
-     * we need to filter by the intersection of the two
-     */
-    producersIds.length > 0 && eservicesIds.length > 0
-      ? {
-          "data.eserviceId": {
-            $in: eservicesIds.filter((eserviceId) =>
-              producerEServiceIds.includes(eserviceId)
-            ),
-          },
-        }
-      : ReadModelRepository.arrayToFilter(
-          [...eservicesIds, ...producerEServiceIds],
-          {
-            "data.eserviceId": {
-              $in: [...eservicesIds, ...producerEServiceIds],
-            },
-          }
-        );
+  const eservicesIdsFilter = match({
+    hasProducersFilter: producersIds.length > 0,
+    hasEServiceFilter: eservicesIds.length > 0,
+  })
+    .returnType<ReadModelFilter<Purpose>>()
+    .with({ hasProducersFilter: true, hasEServiceFilter: true }, () => ({
+      "data.eserviceId": {
+        $in: eservicesIds.filter((eserviceId) =>
+          producerEServicesIds.includes(eserviceId)
+        ),
+      },
+    }))
+    .with({ hasProducersFilter: true }, () => ({
+      "data.eserviceId": {
+        $in: producerEServicesIds,
+      },
+    }))
+    .with({ hasEServiceFilter: true }, () => ({
+      "data.eserviceId": {
+        $in: eservicesIds,
+      },
+    }))
+    .otherwise(() => ({}));
```

Comment:
We do not include the `data.eserviceId` filter ONLY IF both of the producer filter and eservice filter is not being passed